adapters/repos/db/lsmkv/segment_group_compaction.go

```diff
@@ -27,75 +26,143 @@ import (
+// selected for compaction first producing single segment s4s5(3), then s2+s3 producing s2s3(4),
+// then s1+s2s3 producing s1s2s3(5). Until new segment s9(0) will be added to segment group,
+// no next pair will be returned, as all segments will have different levels.
+//
+// If maxSegmentSize is set, estimated total size of compacted segment must not exceed given limit.
+// Only pair of segments having sum of sizes <= maxSegmentSize can be returned by the method. If there
+// exist older segments with same lavel as level of selected pair, level of compacted segment will
+// not be changed to ensure no new segment have higher level than older ones. If there is no segment
+// having same level as selected pair, new segment's level will be incremented.
+// E.g. out of segments: s1(4), s2(3), s3(3), s4(2), s5(2), s6(2), s7(1), s8(0),
+// when s4.size+s5.size > maxSegmentSize, but s5.size+s6.size <= maxSegmentSize, s5+s6 will be selected
+// first for compaction producing segment s5s6(2) (note same level, due to older s2(2)).
+// If s2.size+s3.size <= maxSegmentSize, then s2+s3 producing s2s3(4) (note incremented level,
+// due to no older segment of level 3). If s1.size+s2s3.size <= maxSegmentSize s1s2s3(5) will be produced,
+// if not, no other pair will be returned until new segments will be added to segment group.
+//
+// If compactLeftOverSegments is set, pair of segments of lowest levels, though similar in sizes
+// can be returned if no pair of matching levels will be found. Segment sizes need to be close to each
+// other to prevent merging large segments (GiB) with tiny one (KiB). Level of newly produced segment
+// will be the same as level of larger(left) segment.
+// maxSegmentSize ise respected for pair of leftover segments.
+func (sg *SegmentGroup) findCompactionCandidates() (pair []int, level uint16) {
 	// if true, the parent shard has indicated that it has
 	// entered an immutable state. During this time, the
 	// SegmentGroup should refrain from flushing until its
 	// shard indicates otherwise
 	if sg.isReadyOnly() {
-		return nil
+		return nil, 0
 	}
 
+	sg.maintenanceLock.RLock()
+	defer sg.maintenanceLock.RUnlock()
+
 	// Nothing to compact
 	if len(sg.segments) < 2 {
-		return nil
-	}
-
-	// first determine the lowest level with candidates
-	levels := map[uint16]int{}
-	lowestPairLevel := uint16(math.MaxUint16)
-	lowestLevel := uint16(math.MaxUint16)
-	lowestIndex := -1
-	secondLowestIndex := -1
-	pairExists := false
-
-	for ind, seg := range sg.segments {
-		levels[seg.level]++
-		val := levels[seg.level]
-		if val > 1 {
-			if seg.level < lowestPairLevel {
-				lowestPairLevel = seg.level
-				pairExists = true
+		return nil, 0
+	}
+
+	matchingPairFound := false
+	leftoverPairFound := false
+	var matchingLeftId, leftoverLeftId int
+	var matchingLevel, leftoverLevel uint16
+
+	// as newest segments are prioritized, loop in reverse order
```

Comment:
newest segments are prioritized only in the case no matching pairs has been found, isn't it?

Reply:
if there are multiple pairs with matching levels, the ones with the lowest level (= newest) are selected. If there are multiple pairs with same level, then the oldest ones are selected so compacted segment with incremented level still was oldest.

E.g:
a) s1(**3**), s2(**3**), s3(**2**), s4(**2**), s5(**1**), s6(**1**) => s5+s6 compacted later on to s5s6(2)
b) s1(**3**), s3(**3**), s3(**2**), s4(**2**), s5(**2**), s6(**1**) => s3+s4 compacted later on to s3s4(3)